### PR TITLE
WIP: fixes #692  bootloader module fails when /EFI/Boot exists

### LINKED ISSUE #692: bootloader module fails when /EFI/Boot exists
```
Boost.Python error in job "bootloader".
<class 'subprocess.CalledProcessError'>
Command 'mkdir -p /boot/efi/EFI/boot' returned non-zero exit status 1
Traceback:
  File "/usr/lib/calamares/modules/bootloader/main.py", line 309, in run
    prepare_bootloader(fw_type)

  File "/usr/lib/calamares/modules/bootloader/main.py", line 283, in prepare_bootloader
    install_grub(efi_directory, fw_type)

  File "/usr/lib/calamares/modules/bootloader/main.py", line 231, in install_grub
    check_target_env_call(["mkdir", "-p", efi_boot_directory])

  File "<string>", line 2, in <module>
```
In the current case /boot/efi/EFI/Boot was created by MS Win 10 - the bug is reproducible, one of our Users reported exactly the same bug on  new machine with a pre installed Win10.

### PR DESCRIPTION
**Pull request by agaida**
to make it short - it doesn't help much if one tries to find and process any
EFI related things in the live system. The better approach is to search in
the chroot.

### CODE REVIEW

File: src/modules/bootloader/main.py [hunk 1]
```diff
@@ -193,9 +193,11 @@ def install_grub(efi_directory, fw_type):
     """
     if fw_type == "efi":
         print("Bootloader: grub (efi)")
+        install_path = libcalamares.globalstorage.value("rootMountPoint")
+        install_efi_directory = install_path + efi_directory
 
-        if not os.path.isdir(efi_directory):
-            check_target_env_call(["mkdir", "-p", "{!s}".format(efi_directory)])
+        if not os.path.isdir(install_efi_directory):
+            subprocess.call(["mkdir", "-p", "{!s}".format(install_efi_directory)])
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Could you explain this change? To me it seems like check_target_env_call(efi_directory) and this subprocess.call would do the exact same thing?

File: src/modules/bootloader/main.py [hunk 2]
```diff
@@ -220,24 +222,25 @@ def install_grub(efi_directory, fw_type):
                                "--force"])
 
         # VFAT is weird, see issue CAL-385
-        efi_directory_firmware = os.path.join(efi_directory, "EFI")
-        if os.path.exists(efi_directory_firmware):
-            efi_directory_firmware = vfat_correct_case(efi_directory, "EFI")
+        install_efi_directory_firmware = vfat_correct_case(install_efi_directory, "EFI")
+        if not os.path.exists(install_efi_directory_firmware):
+            subprocess.call(["mkdir", "-p", install_efi_directory_firmware])
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Likewise, why not call check_target_env_call?

File: src/modules/bootloader/main.py [hunk 3]
```diff
@@ -220,24 +222,25 @@ def install_grub(efi_directory, fw_type):
...
-        if os.path.exists(efi_boot_directory):
-            efi_boot_directory = vfat_correct_case(efi_directory_firmware, "boot")
-        else:
-            check_target_env_call(["mkdir", "-p", efi_boot_directory])
+        # there might be several values for the boot directory
+        # most usual they are boot, Boot, BOOT
+
+        install_efi_boot_directory = vfat_correct_case(install_efi_directory_firmware, "boot")
+        if not os.path.exists(install_efi_boot_directory):
+            subprocess.call(["mkdir", "-p", install_efi_boot_directory])
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Likewise.

File: src/modules/bootloader/main.py [hunk 4]
```diff
@@ -220,24 +222,25 @@ def install_grub(efi_directory, fw_type):
...
-        efi_file_target = {"32": os.path.join(efi_boot_directory, "bootia32.efi"),
-                           "64": os.path.join(efi_boot_directory, "bootx64.efi")}
-        check_target_env_call(["cp",
-                               efi_file_source[efi_bitness],
-                               efi_file_target[efi_bitness]])
+        efi_file_source = {"32": os.path.join(install_efi_directory_firmware, efi_bootloader_id, "grubia32.efi"),
+                           "64": os.path.join(install_efi_directory_firmware, efi_bootloader_id, "grubx64.efi")}
+        efi_file_target = {"32": os.path.join(install_efi_boot_directory, "bootia32.efi"),
+                           "64": os.path.join(install_efi_boot_directory, "bootx64.efi")}
+        subprocess.call(["cp",
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Likewise.

Comment (agaida) [inline src/modules/bootloader/main.py]: Thought off, but i see code in calamares that doing it the same way - it's was easier and less changes.

**Comment by agaida** [comment]
@shadeslayer - first i noticed that the local system is checked and if no EFI is found check_target_env_call is used to create the missed dir. So my point is: We have the complete set of well set variables available and have only to add the install path to a new one and use this new variable in the check - these checks are strictly local, even the vfat correction is local and give back a corrected path with the chroot part included. So i decided to fire up an subprocess and use this directly.

**Comment by shadeslayer** [comment]
The part of the PR that makes the check use the target install dirs are correct afaik, I'm only requesting changes on the subprocess.call bits.

After grepping for subprocess.call in the code, there's only one instance of it in src/modules/bootloader/main.py which can probably made to use check_target_env_call too. 

My recommendation is to merge the parts of this PR that do the check on the target install while keeping the check_target_env_call parts that run the commands.

**Comment by tux3l** [comment]
@shadeslayer I think @agaida wanted to keep the code clean an readable, it is true that you can use check_target_env_call but then you check with for example install_efi_directory (absolute path seen from live-system) and you are using efi_directory (absolute path seen from chroot) in check_target_env_call to create the directory. This can lead to some confused people as they do not know that os.path. is checking in your livesystem and not the chroot. 
And, if i see correct, the function vfat_correct_case can only work with absolute paths from your live-system, it will not work with chroot paths like efi_directory. So install_efi_directory contains now a valid path for creating the efi-directory. 
(vfat_correct_case uses os.listdir() which will look for the path in your live-system not the chroot-system)

**Comment by shadeslayer** [comment]
If you really want to do this with the full path, why use subprocess.call at all?

Why not use the Python built in's to create folders.

**Comment by agaida** [comment]
@shadeslayer - very good question, see the disclaimer, i'm a python noob - you are right. I will change it

**Comment by agaida** [comment]
@shadeslayer - thanks for the hint, looks cleaner now and 'works for me'™

**Comment by philmmanjaro** [comment]
@agaida: since you changed the code, pls add also your copyright to it. Thx.

File: src/modules/bootloader/main.py [hunk 5]
```diff
@@ -220,24 +223,21 @@ def install_grub(efi_directory, fw_type):
...
 
-        efi_boot_directory = os.path.join(efi_directory_firmware, "boot")
-        if os.path.exists(efi_boot_directory):
-            efi_boot_directory = vfat_correct_case(efi_directory_firmware, "boot")
-        else:
-            check_target_env_call(["mkdir", "-p", efi_boot_directory])
+        # there might be several values for the boot directory
+        # most usual they are boot, Boot, BOOT
+
+        install_efi_boot_directory = vfat_correct_case(install_efi_directory_firmware, "boot")
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Needs splitting into two lines to keep pep8 happy

File: src/modules/bootloader/main.py [hunk 6]
```diff
@@ -220,24 +223,21 @@ def install_grub(efi_directory, fw_type):
...
 
         # Workaround for some UEFI firmwares
-        efi_file_source = {"32": os.path.join(efi_directory_firmware, efi_bootloader_id, "grubia32.efi"),
-                           "64": os.path.join(efi_directory_firmware, efi_bootloader_id, "grubx64.efi")}
-        efi_file_target = {"32": os.path.join(efi_boot_directory, "bootia32.efi"),
-                           "64": os.path.join(efi_boot_directory, "bootx64.efi")}
-        check_target_env_call(["cp",
-                               efi_file_source[efi_bitness],
-                               efi_file_target[efi_bitness]])
+        efi_file_source = {"32": os.path.join(install_efi_directory_firmware, efi_bootloader_id, "grubia32.efi"),
```
Comment (shadeslayer) [inline src/modules/bootloader/main.py]: Needs splitting to keep pep8 happy

Comment (agaida) [inline src/modules/bootloader/main.py]: :) - will do so

**Comment by shadeslayer** [comment]
LGTM

**Comment by philmmanjaro** [comment]
@shadeslayer: so we merge it?

**Comment by philmmanjaro** [comment]
Code looks fine so far. Any objections?

**Comment by shadeslayer** [comment]
Merged

**Pull request merged**